src/openforms/submissions/tests/form_logic/test_binding.py

```diff
@@ -0,0 +1,375 @@
+from openforms.variables.constants import DataMappingTypes
+from openforms.variables.tests.factories import ServiceFetchConfigurationFactory
+from openforms.variables.validators import HeaderValidator, ValidationError
+
+from ...form_logic import bind
+
+DEFAULT_REQUEST_HEADERS = {
+    "Accept",
+    "Accept-Encoding",
+    "Connection",
+    "Content-Type",
+    "User-Agent",
+}
+
+
+def data_mapping_values() -> st.SearchStrategy[Any]:
+    "Generates values for the value side of openforms.typing.DataMapping"
+    return st.one_of(st.text(), st.integers(), st.floats(), st.dates(), st.datetimes())
+
+
+# the first test still takes too much time
+# decorating given because I don't want to count on test running order
+given = settings(deadline=500)(given)
+
+
+class ServiceFetchConfigVariableBindingTests(SimpleTestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.service = ServiceFactory.build(
+            api_type=APITypes.orc,
+            api_root="https://httpbin.org/",
+            auth_type=AuthTypes.no_auth,
+            oas_file=FileField(
+                from_path=str(Path(__file__).parent.parent / "files" / "openapi.yaml")
+            ),
+        )
+        # prevent parsing the yaml over and over and over
+        cls.service.id = 1  # need pk for __hash__
+        cls.service.build_client = lru_cache(1)(cls.service.build_client)
+
+    @classmethod
+    def tearDownClass(cls):
+        # we did not touch the database, we're Simple
+        # super() insists on calling _remove_databases_failures
+        pass
+
+    @requests_mock.Mocker()
+    def test_it_performs_simple_get(self, m):
+        m.get("https://httpbin.org/get", json={"url": "https://httpbin.org/get"})
+        var = FormVariableFactory.build(
+            service_fetch_configuration=ServiceFetchConfigurationFactory.build(
+                service=self.service,
+                path="get",
+            )
+        )
+
+        value = bind(var, {})
+
+        self.assertEqual(value["url"], "https://httpbin.org/get")
+
+    @given(data_mapping_values())
+    @example("../otherendpoint")
+    def test_it_can_construct_simple_path_parameters(self, field_value):
+        # https://swagger.io/docs/specification/describing-parameters/#path-parameters
+        context = {"seconds": field_value}
+
+        var = FormVariableFactory.build(
+            service_fetch_configuration=ServiceFetchConfigurationFactory.build(
+                service=self.service,
+                path="delay/{seconds}",  # this is not defined as number in the OAS
+            )
+        )
+
+        with requests_mock.Mocker() as m:
+            m.get(requests_mock.ANY)
+            _ = bind(var, context)
+            request = m.last_request
+
+        head, rest = request.url[:26], request.url[26:]
+        self.assertEqual(head, "https://httpbin.org/delay/")
+        # the service should get the value
+        self.assertEqual(unquote(rest), str(field_value))
```

Comment:
I'm sure there's plenty of value in hypothesis... but I really struggle to read/understand this. What is going on here?

I get really annoyed when I have to execute code in my head to understand assertions instead of simply reading what's supposed to happen.

Reply:
Yet it makes sense: the tests are modelling user behaviour. Users can do anything.
![image](https://user-images.githubusercontent.com/325643/219095157-04379709-7dab-4094-9235-0a258134e492.png)
What puzzles me is the complexity of the implementation: it's a webform.